prow/config/config.go

```diff
@@ -1243,25 +1262,105 @@ type Branding struct {
+	OrgRepo string `json:"repo,omitempty"`
+	// Cluster matches against the cluster alias of the build cluster that the
+	// ProwJob is configured to run on. Recall that ProwJobs default to running on
+	// the "default" build cluster if they omit the "cluster" field in config.
+	Cluster string `json:"cluster,omitempty"`
+
+	// Config is the RerunAuthConfig to apply if the filter fields all match the
+	// ProwJob. Note that when multiple entries match a ProwJob the entry with the
+	// highest specification is used.
+	Config *prowapi.RerunAuthConfig `json:"rerun_auth_configs,omitempty"`
+}
+
+func (d *Deck) GetRerunAuthConfig(refs *prowapi.Refs, cluster string) *prowapi.RerunAuthConfig {
+	var config *prowapi.RerunAuthConfig
+
+	var orgRepo string
+	if refs != nil {
+		orgRepo = refs.OrgRepoString()
 	}
 
-	if rerun, exists := rac[fmt.Sprintf("%s/%s", refs.Org, refs.Repo)]; exists {
-		return rerun
+	for _, drac := range d.DefaultRerunAuthConfigs {
+		if drac.matches(orgRepo, cluster) {
+			config = drac.Config
+		}
 	}
 
-	if rerun, exists := rac[refs.Org]; exists {
-		return rerun
+	return config
+}
+
+// defaultRerunAuthMapToSlice converts the old format
+// (map[string]*prowapi.RerunAuthConfig) to the new format
+// ([]*DefaultRerunAuthConfigEntry) or DefaultRerunAuthConfigs.
+func defaultRerunAuthMapToSlice(m map[string]prowapi.RerunAuthConfig) []*DefaultRerunAuthConfigEntry {
+	var entries []*DefaultRerunAuthConfigEntry
+	add := func(repo string, rac prowapi.RerunAuthConfig) {
+		entries = append(entries, &DefaultRerunAuthConfigEntry{
+			OrgRepo: repo,
+			Cluster: "",
+			Config:  &rac,
+		})
+	}
+
+	// Ensure "*" comes first...
+	if rac, exists := m["*"]; exists {
+		add("*", rac)
+	}
+	// then orgs...
+	for key, rac := range m {
+		if key == "*" || strings.Contains(key, "/") {
+			continue
+		}
+		add(key, rac)
+	}
+	// then repos.
+	for key, rac := range m {
+		if key == "*" || !strings.Contains(key, "/") {
+			continue
+		}
+		add(key, rac)
 	}
 
-	return rac["*"]
+	return entries
+}
+
+// FinalizeDefaultRerunAuthConfigs prepares the entries of
+// Deck.DefaultRerunAuthConfigs for use finalizing job config.
+// It parses either d.RerunAuthConfigs or d.DefaultRerunAuthConfigEntries, not both.
+// Old format: map[string]*prowapi.RerunAuthConfig where the key is org,
+//             org/repo, or "*".
+// New format: []*DefaultRerunAuthConfigEntry
+// If the old format is parsed it is converted to the new format, then all
+// filter regexp are compiled.
+func (d *Deck) FinalizeDefaultRerunAuthConfigs() error {
+	mapped, sliced := len(d.RerunAuthConfigs) > 0, len(d.DefaultRerunAuthConfigs) > 0
+
+	// This case should be guarded against by prow config test. Checking here is
+	// for cases where prow config test didn't run
+	if mapped && sliced {
+		return fmt.Errorf("deck.rerun_auth_configs and deck.default_rerun_auth_configs are mutually exclusive, please use one or the other")
+	}
+
+	// Set up DefaultRerunAuthConfigEntries
```

Comment:
Missing period. Can you double-check everywhere else too?

Reply:
^